File: challenges/ttl-store/php/solution/exercise.php
<?php
declare(strict_types=1);

namespace Refresher;

use Closure;

final class TTLStore implements KeyValueStore
{
    private array $entries = [];

    public function __construct(private Closure $clock) {}

    public function put(string $key, string $value, int $ttlMs): void
    {
        $this->entries[$key] = [$value, ($this->clock)() + $ttlMs];
    }

    public function get(string $key): ?string
    {
        if (!array_key_exists($key, $this->entries)) {
            return null;
        }
        [$value, $deadline] = $this->entries[$key];
        if (($this->clock)() >= $deadline) {
            unset($this->entries[$key]);
            return null;
        }
        return $value;
    }

    public function delete(string $key): bool
    {
        if ($this->get($key) === null) {
            return false;
        }
        unset($this->entries[$key]);
        return true;
    }
}

File: challenges/ttl-store/python/solution/exercise.py
"""Reference: lazy expiration with an injected clock."""
from tooling.python.contracts import Clock


class TTLStore:
    def __init__(self, clock: Clock):
        self._clock = clock
        self._entries: dict[str, tuple[str, int]] = {}

    def put(self, key: str, value: str, ttl_ms: int) -> None:
        self._entries[key] = (value, self._clock() + ttl_ms)

    def get(self, key: str) -> str | None:
        entry = self._entries.get(key)
        if entry is None:
            return None
        value, deadline = entry
        if self._clock() >= deadline:
            del self._entries[key]
            return None
        return value

    def delete(self, key: str) -> bool:
        if self.get(key) is None:
            return False
        del self._entries[key]
        return True
